libatomic problems causing unexpected fails

Applications may crash on startup with the following message
```
terminate called after throwing an instance of 'hpx::detail::exception_with_info<hpx::exception>'
  what():  big_boot_barrier has not been created yet: HPX(internal_server_error)
```
upon investigation, it turns out that the following exception is being thrown during initialization/startup
https://github.com/STEllAR-GROUP/hpx/blob/1527f3ba8055f795f701ddf89368f4a24199f79e/libs/full/threadmanager/src/threadmanager.cpp#L322-L329

which in turn causes the runtime destructor to be called, but the runtime has not yet started, so it throws the error that we actually see about the big boot barrier not being started. So the first exception needs to be caught somewhere and displayed, before the stack unwinds and throws the second.

Looking deeper, the problem is being caused by the following check failing during cmake configuration. 
```
-- Looking for __atomic_fetch_add_4 in atomic
-- Looking for __atomic_fetch_add_4 in atomic - not found
```
which in turn means that `HPX_WITH_CXX11_ATOMIC_128BIT` is not set. If I manually run the cmake feature check withoout linking libatomic, we see link errors
```
oryx2:~/build/hpx-master$ g++ ~/src/hpx/cmake/tests/cxx11_std_atomic_128bit.cpp
/usr/bin/ld: /tmp/ccduIFup.o: in function `std::atomic<uint128_type>::store(uint128_type, std::memory_order)':
cxx11_std_atomic_128bit.cpp:(.text._ZNSt6atomicI12uint128_typeE5storeES0_St12memory_order[_ZNSt6atomicI12uint128_typeE5storeES0_St12memory_order]+0x66): undefined reference to `__atomic_store_16'
/usr/bin/ld: /tmp/ccduIFup.o: in function `std::atomic<uint128_type>::load(std::memory_order) const':
cxx11_std_atomic_128bit.cpp:(.text._ZNKSt6atomicI12uint128_typeE4loadESt12memory_order[_ZNKSt6atomicI12uint128_typeE4loadESt12memory_order]+0x40): undefined reference to `__atomic_load_16'
collect2: error: ld returned 1 exit status
```
but if I add `-latomic` all is well.
```
oryx2:~/build/hpx-master$ g++ ~/src/hpx/cmake/tests/cxx11_std_atomic_128bit.cpp -latomic
oryx2:~/build/hpx-master$ <links correctly>
```

I have only run into this problem on my new laptop using gcc 10.2.0  

I'm not quite sure what the correct fix for the atomic problem is.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1784,8 +1784,8 @@ if(NOT WIN32)
   # ############################################################################
   if(NOT MSVC)
     hpx_add_compile_flag_if_available(-pthread)
-    if(HPX_HAVE_LIBATOMIC)
-      target_link_libraries(hpx_base_libraries INTERFACE atomic)
+    if(HPX_CXX11_STD_ATOMIC_LIBRARIES)
+      target_link_libraries(hpx_base_libraries INTERFACE ${HPX_CXX11_STD_ATOMIC_LIBRARIES})
     endif()
   endif()
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1785,7 +1785,9 @@ if(NOT WIN32)
   if(NOT MSVC)
     hpx_add_compile_flag_if_available(-pthread)
     if(HPX_CXX11_STD_ATOMIC_LIBRARIES)
-      target_link_libraries(hpx_base_libraries INTERFACE ${HPX_CXX11_STD_ATOMIC_LIBRARIES})
+      target_link_libraries(
+        hpx_base_libraries INTERFACE ${HPX_CXX11_STD_ATOMIC_LIBRARIES}
+      )
     endif()
   endif()
 

--- a/cmake/HPX_AddConfigTest.cmake
+++ b/cmake/HPX_AddConfigTest.cmake
@@ -280,28 +280,34 @@ endfunction()
 function(hpx_check_for_cxx11_std_atomic)
   # Make sure HPX_HAVE_LIBATOMIC is removed from the cache if necessary
   if(NOT HPX_WITH_CXX11_ATOMIC)
-    unset(HPX_HAVE_LIBATOMIC CACHE)
+    unset(HPX_CXX11_STD_ATOMIC_LIBRARIES CACHE)
   endif()
 
-  if(NOT MSVC)
-    # Sometimes linking against libatomic is required for atomic ops, if the
-    # platform doesn't support lock-free atomics. We know, it's not needed for
-    # MSVC
-    check_library_exists(atomic __atomic_fetch_add_4 "" HPX_HAVE_LIBATOMIC)
-    if(HPX_HAVE_LIBATOMIC)
-      set(HPX_CXX11_STD_ATOMIC_LIBRARIES
-          atomic
-          CACHE BOOL "std::atomics need separate library" FORCE
-      )
-    endif()
-  endif()
-
+  # first see if we can build atomics with no -latomics
   add_hpx_config_test(
     HPX_WITH_CXX11_ATOMIC
     SOURCE cmake/tests/cxx11_std_atomic.cpp
     LIBRARIES ${HPX_CXX11_STD_ATOMIC_LIBRARIES}
     FILE ${ARGN}
   )
+
+  if(NOT MSVC)
+    # Sometimes linking against libatomic is required, if the platform doesn't
+    # support lock-free atomics. We already know that MSVC works
+    if(NOT HPX_WITH_CXX11_ATOMIC)
+      set(HPX_CXX11_STD_ATOMIC_LIBRARIES
+          atomic
+          CACHE BOOL "std::atomics need separate library" FORCE
+      )
+      unset(HPX_WITH_CXX11_ATOMIC CACHE)
+      add_hpx_config_test(
+        HPX_WITH_CXX11_ATOMIC
+        SOURCE cmake/tests/cxx11_std_atomic.cpp
+        LIBRARIES ${HPX_CXX11_STD_ATOMIC_LIBRARIES}
+        FILE ${ARGN}
+      )
+    endif()
+  endif()
 endfunction()
 
 # Separately check for 128 bit atomics
@@ -312,6 +318,23 @@ function(hpx_check_for_cxx11_std_atomic_128bit)
     LIBRARIES ${HPX_CXX11_STD_ATOMIC_LIBRARIES}
     FILE ${ARGN}
   )
+  if(NOT MSVC)
+    # Sometimes linking against libatomic is required, if the platform doesn't
+    # support lock-free atomics. We already know that MSVC works
+    if(NOT HPX_WITH_CXX11_ATOMIC_128BIT)
+      set(HPX_CXX11_STD_ATOMIC_LIBRARIES
+          atomic
+          CACHE BOOL "std::atomics need separate library" FORCE
+      )
+      unset(HPX_WITH_CXX11_ATOMIC_128BIT CACHE)
+      add_hpx_config_test(
+        HPX_WITH_CXX11_ATOMIC_128BIT
+        SOURCE cmake/tests/cxx11_std_atomic_128bit.cpp
+        LIBRARIES ${HPX_CXX11_STD_ATOMIC_LIBRARIES}
+        FILE ${ARGN}
+      )
+    endif()
+  endif()
 endfunction()
 
 # ##############################################################################

--- a/cmake/HPX_AddConfigTest.cmake
+++ b/cmake/HPX_AddConfigTest.cmake
@@ -297,7 +297,7 @@ function(hpx_check_for_cxx11_std_atomic)
     if(NOT HPX_WITH_CXX11_ATOMIC)
       set(HPX_CXX11_STD_ATOMIC_LIBRARIES
           atomic
-          CACHE BOOL "std::atomics need separate library" FORCE
+          CACHE STRING "std::atomics need separate library" FORCE
       )
       unset(HPX_WITH_CXX11_ATOMIC CACHE)
       add_hpx_config_test(
@@ -306,6 +306,9 @@ function(hpx_check_for_cxx11_std_atomic)
         LIBRARIES ${HPX_CXX11_STD_ATOMIC_LIBRARIES}
         FILE ${ARGN}
       )
+      if(NOT HPX_WITH_CXX11_ATOMIC_128BIT)
+        unset(HPX_WITH_CXX11_ATOMIC_128BIT CACHE)
+      endif()
     endif()
   endif()
 endfunction()
@@ -324,7 +327,7 @@ function(hpx_check_for_cxx11_std_atomic_128bit)
     if(NOT HPX_WITH_CXX11_ATOMIC_128BIT)
       set(HPX_CXX11_STD_ATOMIC_LIBRARIES
           atomic
-          CACHE BOOL "std::atomics need separate library" FORCE
+          CACHE STRING "std::atomics need separate library" FORCE
       )
       unset(HPX_WITH_CXX11_ATOMIC_128BIT CACHE)
       add_hpx_config_test(
@@ -333,6 +336,9 @@ function(hpx_check_for_cxx11_std_atomic_128bit)
         LIBRARIES ${HPX_CXX11_STD_ATOMIC_LIBRARIES}
         FILE ${ARGN}
       )
+      if(NOT HPX_WITH_CXX11_ATOMIC_128BIT)
+        unset(HPX_WITH_CXX11_ATOMIC_128BIT CACHE)
+      endif()
     endif()
   endif()
 endfunction()

